scripts/ghutils.py

```diff
@@ -92,5 +111,219 @@ def add_issue_url_to_pr(repo, pull_request_number, issue_key):
+            branch_protection_enabled = True
+            # disable branch protection
+            r = branch_api_request(repo, rc, 'DELETE')
+            r.raise_for_status()
+            logger.debug("Branch protection disabled")
+        elif branch_protection_response.status_code == 404:
+            # rc branch exists without protection, so it doesn't need
+            # to be disabled
+            # TODO: create jira issue about unprotected branch?
+            pass
+        else:
+            # failure retrieving branch protection status
+            branch_protection_response.raise_for_status()
+
+        # Delete branch
+        r = repo._session.request(
+            'DELETE',
+            repo.git_refs_urlt.expand(sha="heads/{}".format(rc)))
+        r.raise_for_status()
+        logger.debug("Branch {} deleted".format(rc))
+
+    mainline_sha = repo.branch(mainline).commit.sha
+    logger.debug("Mainline SHA: {}".format(mainline_sha))
+
+    # create rc branch pointing at head of mainline
+    repo.create_ref("refs/heads/{}".format(rc), mainline_sha)
+    logger.debug("Branch {} created".format(rc))
+
+    # Skeleton branch protection data, used to protect a new branch.
+    protection_data = {
+        "required_status_checks": None,
+        "enforce_admins": True,
+        "required_pull_request_reviews": {
+            "dismissal_restrictions": {},
+            "dismiss_stale_reviews": False,
+            "require_code_owner_reviews": False
+        },
+        "restrictions": None
+    }
+
+    # Incorporate previous branch protection data if the branch was
+    # protected perviously
+    if branch_protection_enabled:
+        stored_bpd = branch_protection_response.json()
+        protection_data.update(stored_bpd)
+        # The github api returns enforce_admins as dict, but requires it to
+        # be sent as a bool.
+        protection_data['enforce_admins'] \
+            = stored_bpd['enforce_admins']['enabled']
+
+    # Enable branch protection
+    r = branch_api_request(repo, rc, 'PUT',
+                           data=json.dumps(protection_data))
+    r.raise_for_status()
+    logger.debug("Branch Protection enabled for branch {}".format(rc))
+
+    # Ensure the rc branch was not updated to anything else while it was
+    # unprotected. Stored mainline_sha is used incase mainline has
+    # moved on since the SHA was acquired.
+    assert mainline_sha == repo.branch(rc).commit.sha
+    logger.debug("rc branch update complete")
+
+
+@cli.command()
+@click.pass_obj
+@click.option(
+    '--version',
+    required=True,
+    help="Symbollic name of Release (eg r14.1.99)"
+)
+@click.option(
+    '--ref',
+    help="Reference to create release from (branch, SHA etc)"
+         " May be omitted if supplied to clone earlier in the chain."
+)
+@click.option(
+    '--bodyfile',
+    help="File containing release message body"
+    # Can't use type=click.File because the file may not exist on startup
```

Comment:
Maybe this whole CLI that does all the steps at the same time should then broken into smaller bits, to then rely on the existing click function and for ease of use?

Reply:
Yeah, That would only be a change to the job, as the python modules can be used independently in their current state.